pkg/util/subnet/subnet_test.go

```diff
@@ -0,0 +1,205 @@
+		wantErr  string
+	}
+
+	for _, tt := range []*test{
+		{
+			name:     "valid",
+			subnetID: "/subscriptions/subscriptionId/resourceGroups/vnetResourceGroup/providers/Microsoft.Network/virtualNetworks/vnet/subnets/subnet",
+			mocks: func(tt *test, subnets *mock_network.MockSubnetsClient) {
+				subnets.EXPECT().
+					CreateOrUpdateAndWait(ctx, "vnetResourceGroup", "vnet", "subnet", mgmtnetwork.Subnet{}).
+					Return(nil)
+			},
+		},
+		{
+			name:     "internal error",
+			subnetID: "/subscriptions/subscriptionId/resourceGroups/vnetResourceGroup/providers/Microsoft.Network/virtualNetworks/vnet/subnets/subnet",
+			mocks: func(tt *test, subnets *mock_network.MockSubnetsClient) {
+				subnets.EXPECT().
+					CreateOrUpdateAndWait(ctx, "vnetResourceGroup", "vnet", "subnet", mgmtnetwork.Subnet{}).
+					Return(fmt.Errorf("random error"))
+			},
+			wantErr: "random error",
+		},
+		{
+			name:     "path error",
+			subnetID: "/invalid/subscriptionId/resourceGroups/vnetResourceGroup/providers/Microsoft.Network/virtualNetworks/vnet/subnets/subnet",
+			wantErr:  "parsing failed for /invalid/subscriptionId/resourceGroups/vnetResourceGroup/providers/Microsoft.Network/virtualNetworks/vnet. Invalid resource Id format",
+		},
+		{
+			name:     "invalid",
+			subnetID: "invalid",
+			wantErr:  `subnet ID "invalid" has incorrect length`,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			controller := gomock.NewController(t)
+			defer controller.Finish()
+
+			subnets := mock_network.NewMockSubnetsClient(controller)
+			if tt.mocks != nil {
+				tt.mocks(tt, subnets)
+			}
+
+			m := &manager{
+				subnets: subnets,
+			}
+
+			err := m.CreateOrUpdate(ctx, tt.subnetID, &mgmtnetwork.Subnet{})
+			if err != nil && err.Error() != tt.wantErr ||
+				err == nil && tt.wantErr != "" {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+func TestNetworkSecurityGroupID(t *testing.T) {
+	oc := &api.OpenShiftCluster{
+		Properties: api.Properties{
+			ClusterProfile: api.ClusterProfile{
+				ResourceGroupID: "/subscriptions/subscriptionId/resourceGroups/clusterResourceGroup",
+			},
+			MasterProfile: api.MasterProfile{
+				SubnetID: "/subscriptions/subscriptionId/resourceGroups/vnetResourceGroup/providers/Microsoft.Network/virtualNetworks/vnet/subnets/master",
+			},
+			WorkerProfiles: []api.WorkerProfile{
+				{
+					SubnetID: "/subscriptions/subscriptionId/resourceGroups/vnetResourceGroup/providers/Microsoft.Network/virtualNetworks/vnet/subnets/worker",
+				},
+			},
+		},
+	}
+
+	for _, tt := range []struct {
+		name      string
+		subnetID  string
+		wantNSGID string
+		wantErr   string
+	}{
+		{
+			name:      "master",
+			subnetID:  "/subscriptions/subscriptionId/resourceGroups/vnetResourceGroup/providers/Microsoft.Network/virtualNetworks/vnet/subnets/master",
+			wantNSGID: "/subscriptions/subscriptionId/resourceGroups/clusterResourceGroup/providers/Microsoft.Network/networkSecurityGroups/aro-controlplane-nsg",
```

Comment:
I noticed that we've got has NSG names hardcoded in `subnet` and in `installed`. Our current tests will not detect the fact that we updated it in one place and failed to do so in other places.

Should we create a medium priority issue to review codebase and replace values like this with constants and\or add tests to cover this?

Reply:
Yes.  A whole pile of these constants are implicitly mandated by the installer, or worse, cluster-api-provider-azure.